File: sem3/map/Lab/3/toylanguage/src/main/java/com/company/Models/Statements/CloseFileStatement.java
package com.company.Models.Statements;

import com.company.Models.*;
import com.company.Models.ProgramState.ProgramState;
import com.company.Utils.MyIDictionary;
import com.company.Utils.Pair;

import java.io.BufferedReader;
import java.io.IOException;

public class CloseFileStatement implements IStatement {

    private Expression expressionFileID;

    public CloseFileStatement(Expression expressionFileID) {
        this.expressionFileID = expressionFileID;
    }

    @Override
    public ProgramState execute(ProgramState state) throws Exception {
        MyIDictionary<String, Integer> symbolTable = state.getSymTable();
        MyIDictionary<Integer, Pair<String, BufferedReader>> fileTable = state.getFileTable();

        int fileID = this.expressionFileID.eval(symbolTable);

        Pair<String, BufferedReader> fileHandler = fileTable.getValue(fileID);

        BufferedReader reader = fileHandler.getSecond();

        try {
            reader.close();
        } catch(IOException e) {
            throw ExceptionFactory.create("file", "Couldn't close the file!");
        }

        fileTable.remove(fileID);

        return state;
    }

    @Override
    public String toString() {
        return "CloseFile(" + this.expressionFileID + ")";
    }
}

File: sem3/map/Lab/toylanguage_C#/ToyLanguage/ToyLanguage/Models/Statements/CloseFileStatement.cs
using System;
using System.IO;
using ToyLanguage.Models.Interfaces;
using ToyLanguage.Models.States;

namespace ToyLanguage.Models.Statements
{
    public class CloseFileStatement : IStatement
    {
        private readonly IExpression _expressionFileID;

        public CloseFileStatement(IExpression expressionFileId)
        {
            _expressionFileID = expressionFileId;
        }


        public ProgramState Execute(ProgramState state)
        {
            int fileID = _expressionFileID.Eval(state);

            StreamReader fileStream = state.FileTable[fileID].Item2;

            try
            {
                fileStream.Close();
            }
            catch (Exception e)
            {
                throw new Exception($"Can\'t close file with ID = {fileID}");
            }

            state.FileTable.Remove(fileID);

            return state;
        }

        public override string ToString()
        {
            return $"CloseFile({_expressionFileID})";
        }
    }
}